keps/sig-network/3836-kube-proxy-improved-ingress-connectivity-reliability/README.md

```diff
@@ -0,0 +1,423 @@
+
+#### Unit tests
+
+Update the Kube-proxy unit tests to include the `healthz` answer for the
+healthcheck server test suite.
+
+#### Integration tests
+
+Kubernetes is mostly tested via unit tests and e2e, not integration, and this is
+not expected to change.
+
+#### e2e tests
+
+- Add E2E tests for connectivity to services on terminating nodes and
+  validate graceful termination of TCP connections.
+
+### Graduation Criteria
+
+#### Alpha
+
+- E2E tests coded before any feature implementation is made which highlights the
+  existing problem.
+- Feature implemented behind a feature flag.
+
+#### Beta
+
+- Document written at https://kubernetes.io/docs/concepts/
+- No issues reported.
+
+#### GA
+
+- No issues reported during two releases.
+
+### Upgrade / Downgrade Strategy
+
+Any upgrade to a version enabling the feature, succeeded by a downgrade to a
+version disabling it, is not expected to be impact ingress in any way, given
+that Kube-proxy is healthy on all cluster nodes. Should Kube-proxy not be
+healthy: then ingress for `eTP:Local` services, might be impacted on an upgrade
+since Kube-proxy would start failing the HC, whereas it wasn't before the
+upgrade. 
+
+### Version Skew Strategy
+
+This doesn't touch load balancer / HC API, so even though an old Kube-proxy
+might talk to a newer control plane, there's no real concern.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [X] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: `KubeProxyDrainingTerminatingNodes`
+  - Components depending on the feature gate: Kube-proxy
+
+###### Does enabling the feature change any default behavior?
+
+Yes. For `eTP:Local` services: Kube-proxy currently just returns the amount of
+endpoints the service has running on the node in question. Kube-proxy will now
+return an answer which also reflects its health AND how many endpoints the
+service has running on the node.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, by resetting the feature gate back.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Behavior will be restored back immediately.
+
+###### Are there any tests for feature enablement/disablement?
+
+- Validate that TCP ingress connections are gracefully terminated when:
+
+  * communicating with endpoints on terminating nodes.  
+
+  * communicating with endpoints not on terminating nodes, but for which the
+    terminating node act as nexthops for traffic.
+
+- Validate that new ingress connections to `eTP:Local` endpoints fail when
+  Kube-proxy on that Node is not healthy.
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+This change is localized to Kube-proxy only. On upgrades Kube-proxy will
```

Comment:
Again, if we change the URL, this does get harder.

Reply:
I will not add changing the URL to this KEP, because of what I mentioned: https://github.com/kubernetes/enhancements/pull/3837#discussion_r1100642321